### PR TITLE
HACK: gs_usb: add termination support

### PR DESCRIPTION
**Pull request by marckleinebudde**
Link: https://github.com/candle-usb/candleLight_fw/pull/108#issuecomment-1250369890

### CODE REVIEW

File: drivers/net/can/usb/gs_usb.c [hunk 1]
```diff
@@ -408,85 +511,30 @@ static void gs_usb_receive_bulk_callback(struct urb *urb)
...
-			if (le32_to_cpu(hf->can_id) & CAN_ERR_FLAG)
-				gs_update_state(dev, cf);
-		}
+	if (hf->echo_id == 0xFFFFFFFF) { /* normal rx */
+		if (gs_usb_receive_can_normal_rx(dev, hf))
+			return;
 
-		netdev->stats.rx_packets++;
-		netdev->stats.rx_bytes += hf->can_dlc;
+	} else if (hf->echo_id == 0xFFFFFFFE) { /* Terminating resistor state */
```
Comment (marckleinebudde) [inline drivers/net/can/usb/gs_usb.c:521]: Please don't abuse the `echo_id`. Please implement `GS_USB_BREQ_GET_TERMINATION`.

File: drivers/net/can/usb/gs_usb.c [hunk 2]
```diff
@@ -1117,6 +1193,10 @@ static struct gs_can *gs_make_candev(unsigned int channel,
 		kfree(bt_const_extended);
 	}
 
+	dev->can.termination_const = gs_termination;
+	dev->can.termination_const_cnt = ARRAY_SIZE(gs_termination);
+	dev->can.do_set_termination = gs_set_termination;
+
```
Comment (marckleinebudde) [inline drivers/net/can/usb/gs_usb.c:1199]: We cannot do this unconditionally...Please set `GS_CAN_FEATURE_TERMINATION` if your hardware supports it.

File: drivers/net/can/usb/gs_usb.c (hunk 1 above)
Comment (Daniel-Trevitz) [reply drivers/net/can/usb/gs_usb.c:521]: Fair enough; This part doesn't work anyways.

File: drivers/net/can/usb/gs_usb.c (hunk 2 above)
Comment (Daniel-Trevitz) [reply drivers/net/can/usb/gs_usb.c:1199]: Took me a minute to rationalize this request... yeah you're right.

**Comment by marckleinebudde** [comment]
See my take on the Linux driver:

https://lore.kernel.org/all/20220918202348.675850-1-mkl@pengutronix.de/

**Comment by Daniel-Trevitz** [comment]
Looks good. Reworking candlelight to match

**Comment by Daniel-Trevitz** [comment]
gs_usb_get_termination does not use the term parameter.

**Comment by marckleinebudde** [comment]
> gs_usb_get_termination does not use the term parameter.

noticed that, working on a v2

**Comment by marckleinebudde** [comment]
Here's the v2 https://lore.kernel.org/all/20220918211802.692405-1-mkl@pengutronix.de. Added some error handling cleanups. The patch series applies to current https://github.com/torvalds/linux/tree/master

**Pull request closed without merging**